File: server/app/schemas/ask.py
"""
问答社区 Pydantic Schema
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class CategoryCreate(BaseModel):
    pid: int = 0
    name: str = Field(..., min_length=1, max_length=100)
    sort_order: int = 0
    is_show: bool = True
    is_show_index: bool = False
    image: str | None = None


class CategoryUpdate(BaseModel):
    id: int
    pid: int | None = None
    name: str | None = None
    sort_order: int | None = None
    is_show: bool | None = None
    is_show_index: bool | None = None
    image: str | None = None


class CategoryOut(BaseModel):
    id: int
    pid: int
    name: str
    sort_order: int
    is_show: bool
    is_show_index: bool
    image: str | None
    level: int
    create_time: datetime | None
    update_time: datetime | None


class QuestionCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    content: str = Field(..., min_length=1)
    image: str | None = None
    cid_list: list[int] | None = []


class QuestionUpdate(BaseModel):
    id: int
    title: str | None = None
    content: str | None = None
    image: str | None = None
    status: str | None = None
    cid_list: list[int] | None = None


class QuestionPageRequest(BaseModel):
    page: int = 1
    limit: int = 10
    keyword: str | None = None
    status: str | None = None
    cid: int | None = None
    member_id: str | None = None
    id_list: list[int] | None = None
    order_column: str | None = "create_time"
    order_direction: Literal["asc", "desc"] | None = "desc"

    @field_validator("order_column")
    @classmethod
    def validate_order_column(cls, v: str | None) -> str | None:
        if v is None:
            return v
        allowed = {
            "id",
            "create_time",
            "update_time",
            "favorite_num",
            "like_num",
            "comment_num",
            "watch_num",
            "answer_num",
        }
        if v not in allowed:
            raise ValueError(f"order_column 不允许: {v}")
        return v


class QuestionOut(BaseModel):
    id: int
    title: str
    content: str
    image: str | None
    member_id: str
    member_name: str | None
    member_avatar: str | None
    status: str
    favorite_num: int
    like_num: int
    comment_num: int
    watch_num: int
    answer_num: int
    is_top: bool
    is_essence: bool
    cid_list: list[int]
    category_list: list[dict]
    create_time: datetime | None
    update_time: datetime | None


class AnswerCreate(BaseModel):
    question_id: int
    content: str = Field(..., min_length=1)


class AnswerUpdate(BaseModel):
    id: int
    content: str | None = None


class AnswerPageRequest(BaseModel):
    page: int = 1
    limit: int = 10
    question_id: int | None = None
    member_id: str | None = None


class AnswerOut(BaseModel):
    id: int
    question_id: int
    content: str
    member_id: str
    member_name: str | None
    member_avatar: str | None
    favorite_num: int
    like_num: int
    comment_num: int
    is_adopted: bool
    is_top: bool
    create_time: datetime | None
    update_time: datetime | None


class CommentCreate(BaseModel):
    target_type: str = Field(..., pattern="^(question|answer)$")
    target_id: int
    content: str = Field(..., min_length=1)
    pid: int = 0

File: client/src/api/ask.ts
/**
 * 问答社区 API
 * 对接后端 /ask/* 端点
 */

import http from '@/utils/request'
import { defaultCache } from '@/utils/requestCache'

export interface AskCategory {
  id: number
  name: string
}

export interface AskQuestion {
  id: number
  title: string
  content: string
  image?: string
  member_id: string
  member_name: string
  member_avatar?: string
  status: string
  favorite_num: number
  like_num: number
  comment_num: number
  watch_num: number
  answer_num: number
  is_top: boolean
  is_essence: boolean
  cid_list: number[]
  category_list: AskCategory[]
  is_like: boolean
  create_time: string
  update_time?: string
}

export interface AskAnswer {
  id: number
  question_id: number
  content: string
  user_id: string
  user_name: string
  user_avatar?: string
  like_num: number
  comment_num: number
  is_adopted: boolean
  is_like: boolean
  create_time: string
}

export interface AskComment {
  id: number
  target_type: 'question' | 'answer'
  target_id: number
  user_id: string
  user_name: string
  content: string
  pid?: number
  create_time: string
}

export const askApi = {
  // 公开问题列表
  listPublic: (params?: { page?: number; limit?: number; keyword?: string; cid?: number; order_column?: string; order_direction?: string }) =>
    http.get('/ask/public-api/list', { params }),

  // 问题详情
  detail: (id: number) => http.get('/ask/public-api', { params: { id } }),

  // 分类列表 - 缓存 5 分钟，分类不常变化
  categories: () =>
    defaultCache.wrap(
      '/ask/category/list',
      () => http.get('/ask/category/list'),
      undefined,
      5 * 60 * 1000
    ),

  // 提问
  create: (data: { title: string; content: string; image?: string; cid_list?: number[]; tags?: string[] }) =>
    http.post('/ask', data),

  // 回答
  answer: (questionId: number, content: string) =>
    http.post('/ask/answer', { question_id: questionId, content }),

  // 回答列表
  answerList: (questionId: number, params?: { page?: number; limit?: number }) =>
    http.get('/ask/answer/list', { params: { question_id: questionId, ...params } }),

  // 点赞
  toggleLike: (targetType: 'question' | 'answer', targetId: number) =>
    http.post('/ask/like', null, { params: { target_type: targetType, target_id: targetId } }),

  // 收藏
  toggleFavorite: (targetType: 'question' | 'answer', targetId: number) =>
    http.post('/ask/favorite', null, { params: { target_type: targetType, target_id: targetId } }),

  // 采纳
  adopt: (answerId: number) => http.post('/ask/answer/adopt', { id: answerId }),

  // 评论
  comment: (data: { target_type: 'question' | 'answer'; target_id: number; content: string; pid?: number }) =>
    http.post('/ask/comment', data),

  // 评论列表
  commentList: (targetType: 'question' | 'answer', targetId: number) =>
    http.get('/ask/comment/list', { params: { target_type: targetType, target_id: targetId } }),
}
